src/components/AmplitudeAnalytics/Trace.tsx

```diff
@@ -0,0 +1,62 @@
+import { createContext, memo, PropsWithChildren, useContext, useEffect, useMemo } from 'react'
+
+import { sendAnalyticsEvent } from '.'
+import { ElementName, EventName, ModalName, PageName, SectionName } from './constants'
+
+export interface ITraceContext {
+  // Highest order context: eg Swap or Explore.
+  page?: PageName
+
+  // Enclosed section name. Can be as wide or narrow as necessary to
+  // provide context.
+  section?: SectionName | ModalName
+
+  // Element name mostly used to identify events sources
+  // Does not need to be unique given the higher order page and section.
+  element?: ElementName
+}
+
+export const TraceContext = createContext<ITraceContext>({})
+
+type TraceProps = {
+  shouldLogImpression?: boolean // whether to log impression on mount
+  eventName?: EventName
+  eventProperties?: Record<string, unknown>
+} & ITraceContext
+
+/**
+ * Sends an analytics event on mount (if shouldLogImpression is set),
+ * and propagates the context to child traces.
+ */
+export const Trace = memo(
+  ({
+    children,
+    shouldLogImpression,
+    page,
+    section,
+    element,
+    eventName,
+    eventProperties,
+  }: PropsWithChildren<TraceProps>) => {
+    const parentTrace = useContext(TraceContext)
+
+    // Component props are destructured to ensure shallow comparison
+    const combinedProps = useMemo(
+      () => ({
+        ...parentTrace,
+        ...Object.fromEntries(Object.entries({ page, section, element }).filter(([_, v]) => v !== undefined)),
+      }),
+      [element, parentTrace, page, section]
+    )
+
+    useEffect(() => {
+      if (shouldLogImpression) {
+        sendAnalyticsEvent(eventName ?? EventName.PAGE_VIEWED, { ...combinedProps, ...eventProperties })
+      }
+    }, [combinedProps, shouldLogImpression, eventName, eventProperties])
```

Comment:
This will not just be sent on mount, but also if combinedProps, shouldLogImpression, eventName, or eventProperties change. If this should _only_ log on mount, then this should be an empty array, with eslint disabled (which is fine and preferable to adding state, so long as you include a comment explaining the intention).

Reply:
sounds good, changed. initially I thought we did want to log on these changes, but after testing with the token selector events it's ok to omit.